src/transformers/models/data2vec/modeling_data2vec_audio.py

```diff
@@ -1356,16 +1356,12 @@ def __init__(self, config, layer_id=0):
         self.kernel = nn.Linear(self.in_conv_dim * self.kernel_size, self.out_conv_dim)
         self.activation = nn.ReLU()
 
-    def forward(self, hidden_states):
-        hidden_states = hidden_states.unsqueeze(1)
-        hidden_states = nn.functional.unfold(
-            hidden_states,
-            (self.kernel_size, self.in_conv_dim),
-            stride=(1, self.in_conv_dim),
-            dilation=(self.dilation, 1),
-        )
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        # for backward compatibility, we keep nn.Linear but call F.conv1d for speed up
+        hidden_states = hidden_states.transpose(1, 2)
+        weight = self.kernel.weight.view(self.out_conv_dim, self.kernel_size, self.in_conv_dim).transpose(1, 2)
+        hidden_states = nn.functional.conv1d(hidden_states, weight, self.kernel.bias, dilation=self.dilation)
         hidden_states = hidden_states.transpose(1, 2)
-        hidden_states = self.kernel(hidden_states)
```

Comment:
cc @younesbelkada as well: currently we have un-optimised TDNN code, where we call `nn.functional.unfold` to do a conv1d, then apply a linear layer `self.kernel`

@gau-nernst has proposed is a solution where we still use the weights of the linear `self.kernel` layer (rather than switching to a conv1d layer which would change the state dict), but then use `nn.functional.conv1d` to do the convolution

=> the benchmarks show there's about a 10% performance gain by doing this

However, we worry that since we don't call `self.kernel.forward`, and LoRA layers for this layer will no longer work for inference. Is this hunch correct?

Reply:
Correct, these changes will make LoRA unusable for the kernel weights on PEFT side as we call `torch.nn.functional.conv1d` instead of the forward of the module itself.
However, usually we only adapt self attention modules (query, key, value) and in rare cases MLP layers of the transformer block. So I am not sure users really adapt `kernel` layers. 
I am not sure if we could make that PR BC (through a config variable maybe?)

cc @pacman100 & @BenjaminBossan as well

Reply:
If this is an edge case that is unlikely to be practically used, I would be fine with breaking backwards compatibility here, given the benefits.

If we are worried about this, however, we could make a check and raise an error/give a warning if a user happens to apply LoRA here. That check could be made here (check if `self.kernel` looks like a LoRA layer) or we could add a check in PEFT if we detect that a user tries to apply LoRA to this specific layer of this specific model.

Btw. this also affects other adapters such as IA³.

Reply:
Awesome - thanks both for your input! It sounds like it's super low likelihood that a user is applying LoRA to these TDNN kernel layers, so I'd be happy proceeding with these changes if it's good for you @gau-nernst @ArthurZucker?

If not, and you'd like to preserve backwards comp for this edge case, I like @BenjaminBossan's idea of checking whether the weights look like a LoRA layer, and triggering a warning if so. There's also the option of having **two** implementations for the TDNN layer:
1. Old, un-optimised variant that is LoRA compatible
2. New, optimised variant that is not LoRA compatible

=> we could use 2 by default, but switch to 1 if the user has applied LoRA to the kernel

Reply:
I'm fine with just warning the user if there are lora weights !

Reply:
Let's do that! Are you happy to proceed with the proposed changes @gau-nernst and throw a warning if the user is using LoRA weights?

Reply:
Sure. I have some ideas about how to detect LoRA
1. Check if the linear layer is a subclass of PEFT's `lora.LoraLayer` (or `lora.Linear`). This requires importing PEFT (we can skip if there is no PEFT installed). This won't be able to detect other LoRA implementations. `torch.compile()` is likely to still work.
2. Check if the linear layer has attribute `lora_A` and/or `lora_B`. This assumes the LoRA implementation uses those names. This approach can potentially detect more LoRA implementations, apart from PEFT. However, iirc, `getattr()` will create a graph break in `torch.compile()`.

What are your thoughts?

Reply:
> However, iirc, `getattr()` will create a graph break in `torch.compile()`.

Interesting, I didn't know that. Would the same happen when using `try: layer.lora_A; except AttributeError: ...`? Because I would prefer the 2nd version, so it would be unfortunate if it cannot work with `torch.compile`. It is extremely unlikely that we would change the attribute name in PEFT so this check would be robust.

Reply:
Fine for me to do number 1 since probably most users will be using PEFT with `transformers`. Checking for weight names `lora_A` and/or `lora_B` is a bit more ad-hoc. So even if it might catch more cases, it's prone to missing / breaking with later versions of LoRA when these vars might change.

Reply:
Sorry for the long overdue update. I added the PEFT LoRA check then realized PEFT LoRA does not even work with Wav2Vec2 😂.

```python
from transformers.models.wav2vec2 import Wav2Vec2Model
from peft import LoraConfig, get_peft_model

model = Wav2Vec2Model.from_pretrained("facebook/wav2vec2-base")

config = LoraConfig(target_modules=["q_proj"])
lora_model = get_peft_model(model, config)
```

Error message

```
/Users/xxx/github/transformers/src/transformers/configuration_utils.py:380: UserWarning: Passing `gradient_checkpointing` to a config initialization is deprecated and will be removed in v5 Transformers. Using `model.gradient_checkpointing_enable()` instead, or if you are using the `Trainer` API, pass `gradient_checkpointing=True` in your `TrainingArguments`.
  warnings.warn(
Traceback (most recent call last):
  File "/Users/xxx/github/transformers/debug.py", line 7, in <module>
    lora_model = get_peft_model(model, config)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/mapping.py", line 103, in get_peft_model
    return PeftModel(model, peft_config, adapter_name=adapter_name)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/peft_model.py", line 119, in __init__
    model = self._prepare_model_for_gradient_checkpointing(model)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/peft_model.py", line 331, in _prepare_model_for_gradient_checkpointing
    model.enable_input_require_grads()
  File "/Users/xxx/github/transformers/src/transformers/modeling_utils.py", line 1333, in enable_input_require_grads
    self._require_grads_hook = self.get_input_embeddings().register_forward_hook(make_inputs_require_grads)
  File "/Users/xxx/github/transformers/src/transformers/modeling_utils.py", line 1352, in get_input_embeddings
    raise NotImplementedError
NotImplementedError
```

where should I open an issue for this? transformers or peft?

Reply:
The error above is due to `gradient_checkpointing=True` in `facebook/wav2vec2-base` config. Setting it to False will resolve the issue. Still, it is indeed a bug i.e. Wav2Vec2 + gradient_checkpointing + PEFT LoRA -> fail. Please direct me where to open an issue for this.

I tested `torch.compile()` and found out that it does not support warning in the `forward()` method.

```python
import torch
import warnings

class DummyModel(torch.nn.Module):
    def forward(self, x):
        warnings.warn("Make a warning")
        return x
    
m = torch.compile(DummyModel(), fullgraph=True)(torch.randn(2, 10))
```

Therefore, having a run-time check and warning will always break `torch.compile` graph.

Reply:
Does this 'bug' happen on `Wav2Vec2ForCTC` or `Wav2Vec2ForXVector` as well? Note that `Wav2Vec2Model` is just the base model, so there's not really any notion of exporting it to PEFT for fine-tuning.

Reply:
It does happen to `Wav2Vec2ForXVector`. I used `Wav2Vec2Model` in the above example to isolate the problem to the base model, to make sure it is not from the XVector components.

Reply:
Okay so the problem here is that Wav2Vec2 doesn't have input embeddings: it works directly with the raw audio waveform and computes positional embeddings directly from the input. Since there is no notion of input embeddings, we don't implement the method `.get_input_embeddings`. This breaks the PEFT LoRA logic, which expects this method to be implemented.

This suggests that Wav2Vec2 is not compatible at all with PEFT LoRA as things currently stand. Would you mind indeed opening an issue on PEFT with the codesnippet you shared, and a copy-and-paste explanation of why it doesn't work?

This means though that we don't need to worry about breaking LoRA for Wav2Vec2, since it doesn't work with it anyway at the moment.

Reply:
The error is due to setting `gradient_checkpointing=True` in Wav2Vec2, which is the default for `facebook/wav2vec2-base`. Other Wav2Vec2 checkpoints, like `facebook/wav2vec2-xls-r-300m`, have `gradient_checkpointing=False`, thus there is no error, and PEFT LoRA still works with them. Overriding `gradient_checkpointing=False` can also make `facebook/wav2vec2-base` works with PEFT LoRA.

Reply:
Ah ok thanks for the clarification! Let's stick to our original strategy then of raising a warning if there are LoRA weights

Reply:
Hey @gau-nernst - would you be interested in finishing this PR? Now that we've agreed on the logic here and you've done a great job at laying down the foundations, it should be quite fast to tidy it up and complete the feature!

Reply:
Hello @sanchit-gandhi. I'm happy to finish this. Do let me know which parts you want to be changed.